Training/Training/intlist.c

```diff
@@ -0,0 +1,188 @@
+   while (current->next != NULL)
+      current = current->next;
+   // make the next of the tail to the newNode
+   current->next = newNode;
+   list->count++;
+   return 0;
+}
+
+int Insert (struct LinkedList* list, int ind, int data) {
+   if (list->isDeleted == true)
+      return E_LIST_DNE;
+   struct Node* newNode = CreateNode (data);
+   // If inserting at the beginning
+   if (ind == 0) {
+      newNode->next = list->head;
+      list->head = newNode;
+      list->count++;
+      return 0;
+   }
+   int count = Count (list);
+   struct Node* current = list->head;
+   if (current == NULL) {
+      free (newNode);
+      return E_INVALID_INDEX;
+   }
+   if (ind >= 0 && ind < count) {
+      for (int i = 0; i < ind - 1; i++)
+         current = current->next;
+   } else { // If the position is out of bounds
+      free (current);
+      return E_INVALID_INDEX;
+   }
+   newNode->next = current->next;
+   current->next = newNode;
+   list->count++;
+   return 0;
+}
+
+int RemoveAt (struct LinkedList* list, int ind) {
+   if (list->isDeleted == true)
+      return E_LIST_DNE;
+   // Base case if linked list is empty
+   if (list->head == NULL)
+      return E_EMPTY_LIST;
+   struct Node* temp = list->head;
+   struct Node* prev = NULL;
+   // Case 1: Head is to be deleted
+   if (ind == 0) {
+      list->head = temp->next;
+      free (temp);
+      list->count--;
+      return 0;
+   }
+   // Case 2: Node to be deleted is in middle
+   // Traverse till given position
+   int count = Count (list);
+   if (ind >= 0 && ind < count) {
+      for (int i = 0; i < ind - 1; i++) {
+         temp = temp->next;
+      }
+      prev = temp;
+      temp = temp->next;
+      prev->next = temp->next;
+      free (temp);
+      list->count--;
+      return 0;
+   } else // If given index is found, delete node
+      return E_INVALID_INDEX;
+}
+
+int Remove (struct LinkedList* list, int data) {
+   if (list->isDeleted == true)
+      return E_LIST_DNE;
+   if (list->head == NULL)
+      return E_EMPTY_LIST;
+   // Store head node
+   struct Node* temp = list->head;
+   struct Node* prev = NULL;
+   // If head node itself holds the data to be deleted
+   if (temp->data == data) {
+      // Changed head
+      list->head = temp->next;
+      free (temp);
+      list->count--;
+      return 0;
+   }
+   // Search for the data to be deleted, keep track of the
+   // previous node as we need to change 'prev->next'
+   while (temp != NULL && temp->data != data) {
+      prev = temp;
+      temp = temp->next;
+   }
+   if (temp != NULL) {
+      prev->next = temp->next;
+      free (temp);
+      list->count--;
+      return 0;
+   } else // If data was not present in linked list
+      return E_INVALID_ELEMENT;
+}
+
+int Count (struct LinkedList* list) {
+   if (list->isDeleted == true) return E_LIST_DNE;
+   return list->count;
+}
+
+int Get (struct LinkedList* list, int ind, int* data) {
+   if (list->isDeleted == true) {
+      *data = 0;

```

Comment:
Preferable to not change the value of data. You are returning the error "List does not exist". The calling program should use the data only if there is no error.